### ISSUE TITLE
SSZ rework - static type hinting, minimal impl. - with POC

### ISSUE THREAD

**Issue by protolambda** [ethereum/eth2.0-specs#1064]
POC implementation here: https://gist.github.com/protolambda/ce2525774c7d1366b7820d4e57d865fa
You can load the POC into Google Colab (click badge on top) to play around with the code.

Credits to @JustinDrake for initial suggestions (Link: https://gist.github.com/protolambda/0cc62589bd4dba209da1b2478d39a419), and later feedback on the non-typed iteration.

Older non-typed iteration of rework: https://gist.github.com/protolambda/4dd66338a2f6214779cadb03c222171a

With this new typed approach, we have a more readable SSZ (ignoring the magic parts) than the previous `minimal_ssz.py`, and we enable static-type hints / linting / checks / IDE suggestions / etc. (as requested by @hwwhww in python call discussion)

I'm switching my efforts back to testing now. I think @JustinDrake can direct getting this into a PR form, and splitting it up. And I hope @hwwhww can review if it fulfills the typing requirements (and maybe iterate on the POC).

Example usage snippet:

```python
class Abc(SSZContainer):
  ab: List[uint64]
  x: uint32

class Foobar(SSZContainer):
  foo: Vector[Abc, SSZLen(5)]

class BeaconThing(SSZContainer):
  a: uint8
  b: Foobar
  baz: BytesN[SSZLen(3)]
  quix: bytes

x = BeaconThing()
x.b.foo[ 0].ab = [0x0123456789abcdef]
x.b.foo[ 0].x = 0x11223344
x.b.foo[4].ab = [0xdeadbeefdeadbeef, 0x3333333333333333]
x.b.foo[4].x = 0x55667788
x.baz = b'\xaa\xbb\xcc'
x.quix = b'\xff\xff'
x.a = 0xcc
print(x.serialize().hex())
print(x.hash_tree_root().hex())

y = BeaconThing()
y.a = 0x77
y.b = Foobar(foo=Vector(Abc(ab=[0x1122334455667788], x=0xaabbccdd), Abc(), Abc(x=0x12345678)))
y.quix = b'\x42\x42\x42\x42\x42\x42'
print(y.serialize().hex())

p = uint8(0xff)
print(serialize(p, uint8).hex())
q = uint16(0xaabb)
print(serialize(q, uint16).hex())
```

Note that the "magic" behind the scenes relies on:
- annotations being present
- generics being used to describe vector/list element types, and vector/bytesN sizes.
- context passing annotations (a dict of `name -> type`, but each type also optionally has a `__args__` with the generic type arguments).
- If no context is available (i.e. not serializing a container), you have to pass the type as an extra argument to `serialize` / `hash_tree_root` / etc. Rare use case, but possible. See example that serializes typed ints.

Performance note:
- integers are not boxed. They are described by `NewType`, which is not execution-time. It is sufficient to just have a reference to the NewType instance however for the SSZ implementation. This way there is no overhead for integers. (Alternatively we could box them at a performance cost, and do overflow/underflow checking?)
- `bytes`/`bytesN` SSZ types are matched together with list/vector, but an extra argument `byte_form` lets you handle them differently (mainly for creating zero values). And the `pack` short-circuits to return the Python `bytes` value quickly.

Readability notes:
- we can probably leaf out the `class BeaconThing(SSZContainer):` for just `BeaconThing`, and have the spec extractor deal with it. (similar with what we do now for `SSZType({...})`
- we may want to rename `SSZLen`, if someone can come up with a better name for a length type argument. (just the integer is not a valid python type)

Also, we can introduce back string based typing if we really want, we can just ignore that the type is unknown (a forward-reference), and then convert a type hint to a string by taking `__forward_arg__` from it. But I rather just use types/annotations/generics everywhere.

**Comment by JustinDrake** [comment]
FWIW, I'm working on cleaning up Proto's POC to limit magic. In particular, looking to remove the SSZ switch magic (so that the only magic that remains is the type magic). Curious to know what people thinl of Proto's latest type magic.

**Comment by dankrad** [comment]
>     * integers are not boxed. They are described by `NewType`, which is not execution-time. It is sufficient to just have a reference to the NewType instance however for the SSZ implementation. This way there is no overhead for integers. (Alternatively we could box them at a performance cost, and do overflow/underflow checking?)

My preference is that in the spec we should just take the hit. This is the reference after all and performance should play a very minimal role in designing it. Not having range checking will limit our tests. Implementations can then go with unboxed versions and show that they will still conform to the spec with all the range checking included.

> Also, we can introduce back string based typing if we really want
I'm against this. I never understood the idea of why some types should be strings. Also, it means that the SSZ types cannot be used literally as the type names when implementing, which seems like a disadvantage to me. Finally, I don't think they were consistent ("bytes1" being a basic type and "bytes2" not -- what were they even trying to indicate?)

**Comment by hwwhww** [comment]
I tried to run it with `mypy` check. Currently, it’s broken, but indeed it *seems* more type-hinting-friendly than `SSZType` one.

Some known issues:
1. Some attribute setting/overridden is not acceptable for `mypy`. `uint8.byte_len = 1` can be written as `setattr(uint8, ‘byte_len’, 1)` to make `mypy` happy. (Maybe there’s other solution)
2. `SSZLenAny = type(“SSZLenAny”, (), {})` would cause error `error: Invalid type "proto_ssz.proto_ssz.SSZLenAny"` when checking `L = TypeVar(“L”, bound=SSZLenAny)`. Can it be just an empty class?
    ```python
    class SSZLenAny:
        pass
    ```
4. The basic types looks fine (although needs some adjustment), but `typ = type(“SSZLen_%d” % length, (SSZLenAny,), {})` is defined in runtime.
5. Some function returning types are missing 

---

Some thoughts:
1. I like the annotations! ❤️
2. Generally, type hinting supporting would be a win for me.
3. However, even if we don't take SSZ type magic as part of the spec, IMO it's still not readable/understandable enough to replace the English spec (which I recall it's the goal that [@JustinDrake mentioned](https://github.com/ethereum/eth2.0-specs/issues/859#issuecomment-484310908)). May need more refactorings to achieve that; but also, I can't tell if it's worth or not right now.

**Comment by protolambda** [comment]
@JustinDrake 
> remove the SSZ switch magic

I wish Python had strong type/object pattern matching like Scala, but we're stuck with if-statements. These matchers that output the necessary data, and managed dict of these, with lambdas receiving the matched data, is the best emulation of Scala pattern matching I could think of.
Alternatively, we use some decorator magic to define encoding functions for each type, but it will be more verbose I think. Or less, if there will be a custom spec build script for this.

@dankrad 
> My preference is that in the spec we should just take the hit

I can see that, but prefer not to, as we've to deal with tests and generators running on hunderds of thousands of these integers. If we start boxing them, I think we are stuck with the minimal configuration forever. It would be nice if you can provide a POC with bound-checker integer boxing, so we can see the benefits / performance hit better. But focus on other typing first :)

> I'm against this. I never understood the idea of why some types should be strings.

More or less on same page here. But do want to see some type declaration standardization at some point, similar to ABI, so we can pass around SSZ types in more dynamic contexts. Typing is great for exec. spec safety however.

@hwwhww 

> 1. Some attribute setting/overridden is not acceptable

Forgot about that. Just wanted a way to attach a length in integer form to the type. I don't want to extract it from the type name every time. `setattr` would work, or maybe declare a static function returning the length from the closure?

> 2. SSZLenAny

`type(“SSZLenAny”, (), {})` should be perfectly valid Python. But yes, may just want to declare an empty class instead. The idea here is that we create type bounds for the `L` type paremeter used in the generics. We can't declare a type for every single length (every integer), but we can limit it to a subclass of some type we control. And then `SSZLen` helps create this type.

> `typ = type(“SSZLen_%d” % length, (SSZLenAny,), {}) is defined in runtime.`

Yes, see above, can't declare types for every integer. The other way would be to parametrize the type creation of `Vector` and `BytesN`, but then those become dynamicly defined types, like `SSZLen`. Better just limit it to root of the problem. Given that the type is bound to `SSZLenAny`, we can at least know on type-checking time that a vector has a certain length, and detect inequality between two types of vectors with the same element type, but different lengths.

Alternatively we use strings for length parameters (doesn't please Dankrad) as types, which is valid python (have them be forward-type-references), and never actually declare them. Not sure what mypi would act like there. It's a rabbit hole to attach arbitrary numbers to *static* typing in Python, yay for Go :see_no_evil: 
Aside from the brevity you get from Python sometimes, and power when *defining* tests, I see more value in Go and Rust for *some* parts of the spec like type declaration. (serialization/merkleization, or anything with type interpretation, would be best for languages with pattern matching however).

>  not readable/understandable enough to replace the English spec

Closer than previous string-based if-statement `minimal_ssz.py` approach, but not quite like the spec either, even if you hide the unimportant typing magic parts from the reader. Personally don't see SSZ splitting up into merkleization and serialization anytime soon.

It's much more straightforward if you can:
- just call `serialize` for basic values when packing an array as part of merkleization
- just define what is `fixed_size` based on serialization primitives/helpers
- keep it all encapsulated from any part of the spec:
  - to make it replaceable
  - easy to black-box as a spec reader
  - use it as a clean dependency in other parts of the spec, e.g. validator spec, phase 1+ things, etc.

Hope the typing idea can be picked up by someone, I am moving focus back to test-generation (a problem blocking many more people), and will stop commenting here. :wave:

**Comment by dankrad** [comment]
>     3\. However, even if we don't take SSZ _type_ magic as part of the spec, IMO it's still not readable/understandable enough to replace the English spec (which I recall it's the goal that [@JustinDrake mentioned](https://github.com/ethereum/eth2.0-specs/issues/859#issuecomment-484310908)). May need more refactorings to achieve that; but also, I can't tell if it's worth or not right now.

I imagine that the "switch" statements in the SSZ minimal implementation would become some kind of tables in the spec. Some good work on encode_series might make it more readable. But I still see the need for lots of English words around it :P

I find the types much more readable than the current json-like notation for containers we use in the spec though! @hwwhww I'm a big fan of having better type hinting, it would be great if we can get it to work properly with mypy!

BTW, since we are now changing naming conventions anyway: Can we capitalize all SSZ types, or if we want to keep the basic types lowercase to distinguish them, ensure it's consistent (e.g. Bytes should be uppercase).

**Comment by hwwhww** [comment]
@protolambda 
I'm happy to dig in mypy-compatibility. But I'd say after this PoC is on a branch and pass the SSZ tests so we could more confident to iterate it. :)

**Comment by protolambda** [comment]
@hwwhww Great, I'll put this work in a branch some time later today, if we're through the prototyping phase of SSZ spec typing.
@dankrad Tables sounds great, but some of the cells may be too large for readability, and it would need special code to extract and build.

**Comment by protolambda** [comment]
See PR #1077, let's continue discussion there.

**Comment by protolambda** [comment]
#1077 was merged, and #1180 takes typing even farther (and improves the style/cleanliness too). Closing this.
